README.md

```diff
@@ -1,11 +1,458 @@
+This library is for checking that software
+[contracts](https://en.wikipedia.org/wiki/Design_by_contract) are
+upheld.  In C++ these checks can be especially important for safety
+because failure to satisfy contracts typically leads to [undefined
+behavior](https://en.wikipedia.org/wiki/Undefined_behavior), which can
+manifest as crashes, data loss, and security vulnerabilities.  This
+library largely improves upon the standard
+[`assert`](https://en.cppreference.com/w/cpp/error/assert) macro.
+
+## Design by Contract
+
+[Design by Contract](https://en.wikipedia.org/wiki/Design_by_contract)
+is *the* industry-standard way to describe the requirements and
+guarantees of any software component.  It is based on three concepts:
+
+- **Preconditions**: what the caller of a function must ensure for the
+  function to behave as documented.  A precondition violation
+  indicates a bug in the caller.
+
+- **Postconditions** describe a function's side-effects and return
+  value. Postconditions need not be upheld if the function reports an
+  error (such as memory exhaustion), or if preconditions were
+  violated.  Otherwise, a postcondition violation indicates
+  a bug in the callee.
+
+- **Invariants**: conditions that always hold at some point in the
+  code.  The most common and useful kind of invariants are **class
+  invariants**, which hold at any point where an instance can be
+  inspected from outside the class.
+
+A function's specification must at least describe its preconditions
+and postconditions, and the specification of a class must describe its
+publicly-visible invariants.  Additionally describing these conditions
+in code and checking them at runtime can be  a powerful way to catch
+bugs early and prevent their damaging effects.
+
+## Basic C++ Usage
+
+This is a header-only library.  To use it from C++, simply put the
+`include` directory of this repository in your `#include` path, and
+`#include <adobe/contract_checks.hpp>`.
+
+```c++
+#include <adobe/contract_checks.hpp>
+```
+
+Every executable needs a single contract violation handler that
+determines the program's behavior when a violation is detected.  A
+good starting point is provided by a macro that you can expand in a
+source file such as the one containing your `main` function.
+
+```c++
+ADOBE_DEFAULT_CONTRACT_VIOLATION_HANDLER() // no semicolon
+```
+
+There are three primary macros used to check contracts, each with two
+forms: `ADOBE_PRECONDITION`, `ADOBE_POSTCONDITION`, and
+`ADOBE_INVARIANT`. Each has one required argument and one optional
+argument:
+
+- `condition`: an expression convertible to `bool`; if `false`, the
+  violation handler is invoked.
+- `message`: an expression convertible to `const char*` pointing to a
+  [null-terminated](https://en.cppreference.com/w/cpp/string/byte)
+  message that is additionally passed to the violation handler. The
+  default value is the empty string, `""`.
```

Comment:
Is `nullptr` allowed for the message?

Reply:
Does `nullptr` point to a null-terminated byte string?  I think not.

> In a macro I believe we can force a string literal by using "" message.
I don't understand what you're suggesting.

Reply:
@sean-parent I get it now.  But why would we want to do that?

Reply:
By ensuring a string literal, you don't have to deal with null and lifetime if it escapes (on the down side, you can't build a message at runtime).

Reply:
It can only escape if you return or throw from your handler.  Are you sure it's worth it?  People can do any number of other bad things in their handlers too.